### PR TITLE
[NSRegularExpression] NSCoding implementation

### PR DESCRIPTION
**Pull request by naithar**

### CODE REVIEW

File: Foundation/NSRegularExpression.swift [hunk 1]
```diff
@@ -39,11 +39,39 @@ open class NSRegularExpression: NSObject, NSCopying, NSCoding {
     }
     
     open func encode(with aCoder: NSCoder) {
-        NSUnimplemented()
+        guard aCoder.allowsKeyedCoding else {
+            preconditionFailure("Unkeyed coding is unsupported.")
+        }
+        
+        aCoder.encode(self.pattern._nsObject, forKey: "NS.pattern")
```
Comment (alblue) [inline Foundation/NSRegularExpression.swift]: Shouldn't this be NSPattern, and not NS.pattern?

File: Foundation/NSRegularExpression.swift [hunk 2]
```diff
@@ -39,11 +39,39 @@ open class NSRegularExpression: NSObject, NSCopying, NSCoding {
...
+        aCoder.encode(self.options.rawValue._bridgeToObjectiveC(), forKey: "NS.options")
```
Comment (alblue) [inline Foundation/NSRegularExpression.swift]: I think this should be NSOptions as well ...

File: Foundation/NSRegularExpression.swift [hunk 3]
```diff
@@ -39,11 +39,39 @@ open class NSRegularExpression: NSObject, NSCopying, NSCoding {
...
+        
+        do {
+            try self.init(pattern: pattern._swiftObject, options: Options(rawValue: options.uintValue))
+        } catch {
+            return nil
+        }
+    }
+    
+    open override func isEqual(_ object: Any?) -> Bool {
+        if let other = object as? NSRegularExpression {
```
Comment (alblue) [inline Foundation/NSRegularExpression.swift]: Would this be more swifty if it was 'guard let other else { return false } and then have the return self == other at the end of the function?

Comment (naithar) [inline Foundation/NSRegularExpression.swift]: That was the way i did the checks on #708 at first. If @parkera doesn't mind, i'll switch first `if` with `guard`.

Comment (parkera) [inline Foundation/NSRegularExpression.swift]: Either way is fine; `guard` seems good to me.

**Comment by naithar** [comment]
@parkera I've made requested changes, please review.

**Comment by alblue** [comment]
@swift-ci please test

**Comment by naithar** [comment]
@parkera can we merge this?

**Comment by parkera** [comment]
@swift-ci test and merge

**Pull request merged**